src/runtime_src/core/common/info_memory.cpp

```diff
@@ -411,4 +413,186 @@ xclbin_info(const xrt_core::device * device)
+  boost::property_tree::ptree ptSt_arr;
+  for(auto& str : bit_set)
+    ptSt_arr.push_back(std::make_pair("", boost::property_tree::ptree(str)));
+
+  if (!ptSt_arr.empty())
+    pt.add_child( std::string("bits_set"), ptSt_arr);
+
+  return pt;
+}
+
+static void
+schedulerUpdateStat(xrt_core::device *device)
+{
+  try {
+    // lock xclbin
+    std::string xclbin_uuid = xrt_core::device_query<xq::xclbin_uuid>(device);
+    // dont open a context if xclbin_uuid is empty
+    if(xclbin_uuid.empty())
+	    return;
+    auto uuid = xrt::uuid(xclbin_uuid);
+    device->open_context(uuid.get(), std::numeric_limits<unsigned int>::max(), true);
+    auto at_exit = [] (auto device, auto uuid) { device->close_context(uuid.get(), std::numeric_limits<unsigned int>::max()); };
+    xrt_core::scope_guard<std::function<void()>> g(std::bind(at_exit, device, uuid));
+
+    device->update_scheduler_status();
+  }
+  catch (const std::exception&) {
+    // xclbin_lock failed, safe to ignore
+  }
+}
+
+int 
+getPSKernels(std::vector<ps_kernel_data> &psKernels, const xrt_core::device *device)
+{
+  try {
+    std::vector<char> buf = xrt_core::device_query<xq::ps_kernel>(device);
+    if (buf.empty())
+      return 0;
+    const ps_kernel_node *map = reinterpret_cast<ps_kernel_node *>(buf.data());
+    if(map->pkn_count < 0)
+      return -EINVAL;
+
+    for (unsigned int i = 0; i < map->pkn_count; i++)
+      psKernels.emplace_back(map->pkn_data[i]);
+  }
+  catch (const xq::no_such_key&) {
+    // Ignoring if not available: Edge Case
+  }
+
+  return 0;
+}
+
+boost::property_tree::ptree
+populate_cus(const xrt_core::device *device)
+{
+  schedulerUpdateStat(const_cast<xrt_core::device *>(device));
+
+  boost::property_tree::ptree pt;
+  using cu_data_type = xq::kds_cu_info::data_type;
+  using scu_data_type = xq::kds_scu_info::data_type;
+  std::vector<cu_data_type> cu_stats;
+  std::vector<scu_data_type> scu_stats;
+  boost::property_tree::ptree ptree;
+  try {
+    std::string uuid = xrt::uuid(xrt_core::device_query<xq::xclbin_uuid>(device)).to_string();
+    boost::algorithm::to_upper(uuid);
+    ptree.put("xclbin_uuid", uuid);
+  } catch (xq::exception&) {  }
+
+  try {
+    cu_stats  = xrt_core::device_query<xq::kds_cu_info>(device);
+    scu_stats = xrt_core::device_query<xq::kds_scu_info>(device);
+  }
+  catch (const xq::no_such_key&) {
+    // Ignoring if not available: Edge Case
+  }
+  catch (const std::exception& ex) {
+    ptree.put("error_msg", ex.what());
+    return ptree;
+  }
+
+  for (auto& stat : cu_stats) {
+    boost::property_tree::ptree ptCu;
+    ptCu.put( "name",           stat.name);
+    ptCu.put( "base_address",   boost::str(boost::format("0x%x") % stat.base_addr));
+    ptCu.put( "usage",          stat.usages);
+    ptCu.put( "type", enum_to_str(cu_type::PL));
+    ptCu.add_child( std::string("status"),	get_cu_status(stat.status));
+    pt.push_back(std::make_pair("", ptCu));
+  }
+
+  std::vector<ps_kernel_data> psKernels;
```

Comment:
no camelCase, but see comment on return value of get_ps_kernel, so this line should be changed to something like:

auto ps_kernels = get_ps_kernels(...);

I see the complication in save ps_kernels inside a try and using it outside. In general this function is too long, it would be better to separate pl and ps and run inside appropriate try/catch